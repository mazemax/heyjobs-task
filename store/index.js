import { createStore, applyMiddleware, combineReducers } from 'redux';
import thunkMiddleware from 'redux-thunk';
import { composeWithDevTools } from 'redux-devtools-extension';

import homeReducer from './home/reducer';

const reducer = combineReducers({
    home: homeReducer
});

/**
 * Initialize Store
 * @param preloadedState
 * @returns {Store, <AnyAction>}
 */
export const initializeStore = (preloadedState) => {
    return createStore(
        reducer,
        preloadedState,
        composeWithDevTools(applyMiddleware(thunkMiddleware))
    )
};