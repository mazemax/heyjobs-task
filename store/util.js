import { isEmpty } from 'lodash';

/**
 * Creates an action
 * @param type
 * @param argNames
 * @returns {function(...[*]): {type: *}}
 */
export function makeActionCreator(type, ...argNames) {
    return function(...args) {
        const action = { type };
        argNames = isEmpty(argNames) ? ['payload'] : argNames;
        argNames.forEach((arg, index) => {
            action[argNames[index]] = args[index];
        });
        return action;
    };
}

/**
 * Creates a reducer
 * @param initialState
 * @param handlers
 * @returns {reducer}
 */
export function createReducer(initialState, handlers) {
    return function reducer(state = initialState, action = {}) {
        if (Object.prototype.hasOwnProperty.call(handlers, action.type)) {
            return handlers[action.type](state, action);
        } else {
            return state;
        }
    };
}
