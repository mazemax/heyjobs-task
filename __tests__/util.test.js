/* eslint-env jest */

import React from 'react';
import { makeActionCreator, createReducer } from "../store/util";

describe('Testing Util functions', () => {
    test('When a new action is created', async () => {
        const ACTION_TEST = 'ACTION_TEST';
        const actionFn = makeActionCreator(ACTION_TEST, ['payload']);

        let action = actionFn('argName1');
        expect(action).toMatchSnapshot();
    });

    test('When a new reducer is created', async () => {
        const ACTION_TEST = 'ACTION_TEST';
        const initialState = {
            argName1: 'argName1'
        };
        const actionObject = {
            type: "ACTION_TEST",
            payload: "argName2"
        };

        let reducerFn = createReducer(initialState, {
            [ACTION_TEST]: (state, action) => {
                return {
                    ...state,
                    argName1: action.payload.argName1
                };
            }
        });

        let reducerResult = reducerFn.call(initialState, actionObject);
        expect(reducerResult).toMatchSnapshot();
    });
});