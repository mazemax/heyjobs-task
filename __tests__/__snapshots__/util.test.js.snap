// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Util functions When a new action is created 1`] = `
Object {
  "payload": "argName1",
  "type": "ACTION_TEST",
}
`;

exports[`Testing Util functions When a new reducer is created 1`] = `
Object {
  "payload": "argName2",
  "type": "ACTION_TEST",
}
`;
